pkg/loki/config_wrapper.go

```diff
@@ -96,85 +99,141 @@ func (c *ConfigWrapper) ApplyDynamicConfig() cfg.Source {
 	}
 }
 
-// applyIngesterRingConfig will use whatever config is setup for the ingester ring and use it everywhere else
-// we have a ring configured. The reason for centralizing on the ingester ring as this is been set in basically
-// all of our provided config files for all of time, usually set to `inmemory` for all the single binary Loki's
-// and is the most central ring config for Loki.
+// applyDynamicRingConfigs checks the current config and, depending on the given values, reuse ring configs accordingly.
 //
-// If the ingester ring has its interface names sets to a value equal to the default (["eth0", en0"]), it will try to append
-// the loopback interface at the end of it.
-func applyIngesterRingConfig(r *ConfigWrapper, defaults *ConfigWrapper) error {
-	if reflect.DeepEqual(r.Ingester.LifecyclerConfig.InfNames, defaults.Ingester.LifecyclerConfig.InfNames) {
-		appendLoopbackInterface(r)
+// 1. Gives preference to any explicit ring config set. For instance, if the user explicitly configures Distributor's ring,
+// that config will prevail. This rule is enforced by the fact that the config file and command line args are parsed
+// again after the dynamic config has been applied, so will take higher precedence.
+// 2. If no explicit ring config is set, use the common ring configured if provided.
+// 3. If no common ring was provided, use the memberlist config if provided.
+// 4. If no common ring or memberlist were provided, use the ingester's ring configuration.
+
+// When using the ingester or common ring config, the loopback interface will be appended to the end of
+// the list of default interface names
+func applyDynamicRingConfigs(r, defaults *ConfigWrapper) {
+	if reflect.DeepEqual(r.Common.Ring, defaults.Common.Ring) {
+		// common ring not set, so use memberlist for all rings if set
+		if len(r.MemberlistKV.JoinMembers) > 0 {
+			applyMemberlistConfig(r)
+		} else {
+			// neither common ring nor memberlist set, use ingester ring configuration for all rings
+			useIngesterRingConfig(r, defaults)
+		}
+	} else {
+		// common ring is provided, use that for all rings
+		useCommonRingConfig(r, defaults)
 	}
 
-	lc := r.Ingester.LifecyclerConfig
-	rc := r.Ingester.LifecyclerConfig.RingConfig
-	s := rc.KVStore.Store
-	sc := r.Ingester.LifecyclerConfig.RingConfig.KVStore.StoreConfig
+}
 
-	f, err := tokensFile(r, "ingester.tokens")
-	if err != nil {
-		return err
-	}
-	r.Ingester.LifecyclerConfig.TokensFilePath = f
+func useCommonRingConfig(r, defaults *ConfigWrapper) {
+	// if the default ingester config is detected, then it's safe to override with the common config
+	shouldOverrideIngester := reflect.DeepEqual(r.Ingester.LifecyclerConfig, defaults.Ingester.LifecyclerConfig)
+
+	// append the loopback after checking if the default is enabled, as it mutates the current
+	appendLoopbackInterface(r, defaults)
+
+	numTokens := 128 //this is the default used by the ingester and ruler
```

Comment:
@owen-d per your comment on @DylanGuedes's original PR, do you think we should use 512 here instead?

Reply:
I don't think we should be overwriting these at all. Extra fields (i.e. num_tokens) can still be reconfigured by setting the ingester ring tokens manually, but I don't think we should be overriding them here.

Reply:
ok, so just leave that one blank for now, I'm in agreement with that.